pkg/secrets/watch.go

```diff
@@ -0,0 +1,266 @@
+				watcher.close()
+				return
+			}
+		}
+	}()
+
+	return watcher, nil
+}
+
+func (w *secretWatcher) update(logger log.Logger, e watch.Event) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	switch e.Type {
+	case watch.Modified, watch.Added:
+		secret := e.Object.(*corev1.Secret)
+		w.s = secret
+	case watch.Deleted:
+		w.s = nil
+	case watch.Bookmark:
+		// Disabled explicitly when creating the watch interface.
+	case watch.Error:
+		//nolint:errcheck
+		logger.Log("msg", "watch error event", "namespace", w.s.Namespace, "name", w.s.Name)
+	}
+}
+
+func (w *secretWatcher) secret(config *KubernetesSecretConfig) Secret {
+	fn := SecretFn(func(_ context.Context) (string, error) {
+		w.mu.Lock()
+		defer w.mu.Unlock()
+		if w.s == nil {
+			return "", fmt.Errorf("secret %s/%s not found", config.Namespace, config.Name)
+		}
+		return getKey(w.s, config.Key)
+	})
+	return &fn
+}
+
+func (w *secretWatcher) start(ctx context.Context, client kubernetes.Interface, config *KubernetesSecretConfig) error {
+	var err error
+	w.w, err = client.CoreV1().Secrets(config.Namespace).Watch(ctx, metav1.ListOptions{
+		FieldSelector:       fields.OneTermEqualSelector(metav1.ObjectNameField, config.Name).String(),
+		AllowWatchBookmarks: false,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to watch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+
+	// We could wait for the first watch event, but it doesn't notify us if the resource doesn't exist.
+	w.s, err = client.CoreV1().Secrets(config.Namespace).Get(ctx, config.Name, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("unable to fetch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+	return nil
+}
+
+// restart returns true and restarts the watch if there are any open secrets.
+func (w *secretWatcher) restart(ctx context.Context, client kubernetes.Interface, config *KubernetesSecretConfig) (bool, error) {
+	// Check in case the channel cancelled intentionally.
+	if w.done {
+		w.mu.Lock()
+		defer w.mu.Unlock()
+		w.s = nil
+		return true, nil
+	}
+
+	// Pseudo-arbitrarily jitter the length of the most common scrape interval.
+	time.Sleep(1*time.Second + time.Second*time.Duration(rand.Intn(30)))
+
+	// Lock the watcher so it doesn't cancel before we restart.
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	// Check again in case the watcher cancelled while we were waiting for the mutex.
+	if w.done {
+		w.s = nil
+		return true, nil
+	}
+
+	if err := w.start(ctx, client, config); err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
+
+func (w *secretWatcher) close() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.w.Stop()
+	w.s = nil
+}
+
+type watchProvider struct {
+	ctx                context.Context
+	client             kubernetes.Interface
+	secretKeyToWatcher map[string]*secretWatcher
+	logger             log.Logger
+}
+
+func newWatchProvider(ctx context.Context, logger log.Logger, client kubernetes.Interface) (*watchProvider, error) {
```

Comment:
suggestion: this doesn't ever return an error. remove that return arg. Or why not just call directly in the `newProvider` call above? i.e.
```go
func (c *WatchSPConfig) newProvider(ctx context.Context, opts ProviderOptions) (*watchProvider, error) {
	client, err := c.ClientConfig.client()
	if err != nil {
		return nil, err
	}
	return &watchProvider{
		ctx:                ctx,
		client:             client,
		secretKeyToWatcher: map[string]*secretWatcher{},
		logger:             logger,
	}, nil
}
```

Reply:
`newWatchProvider` is a separate method because it is used by the unit tests, where it is initialized using a fake Kubernetes API client.